plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiClassBuilderInterceptor.kt

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.jvm.abi
+
+import com.intellij.psi.PsiElement
+import org.jetbrains.kotlin.codegen.ClassBuilder
+import org.jetbrains.kotlin.codegen.ClassBuilderFactory
+import org.jetbrains.kotlin.codegen.DelegatingClassBuilder
+import org.jetbrains.kotlin.codegen.DelegatingClassBuilderFactory
+import org.jetbrains.kotlin.codegen.extensions.ClassBuilderInterceptorExtension
+import org.jetbrains.kotlin.codegen.inline.coroutines.FOR_INLINE_SUFFIX
+import org.jetbrains.kotlin.descriptors.DescriptorVisibilities
+import org.jetbrains.kotlin.descriptors.FunctionDescriptor
+import org.jetbrains.kotlin.descriptors.MemberDescriptor
+import org.jetbrains.kotlin.diagnostics.DiagnosticSink
+import org.jetbrains.kotlin.load.java.JvmAnnotationNames
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.jvm.diagnostics.JvmDeclarationOrigin
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
+import org.jetbrains.org.objectweb.asm.AnnotationVisitor
+import org.jetbrains.org.objectweb.asm.MethodVisitor
+import org.jetbrains.org.objectweb.asm.Opcodes
+import org.jetbrains.org.objectweb.asm.commons.Method
+
+enum class AbiMethodInfo {
+    KEEP,
+    STRIP,
+}
+
+sealed class AbiClassInfo {
+    object Public : AbiClassInfo()
+    class Stripped(val methodInfo: Map<Method, AbiMethodInfo>) : AbiClassInfo()
+}
+
+/**
+ * Record ABI information for all classes in the current compilation unit.
+ *
+ * This needs to be a `ClassBuilderInterceptor`, since we need the descriptors
+ * in order to know which methods can be safely stripped from the output.
+ * On the other hand we cannot produce any output in this pass, since we need
+ * to know which classes are part of the public ABI in order to produce the
+ * correct `InnerClasses` attributes.
+ *
+ * ---
+ *
+ * Classes which are private, local, or anonymous can be stripped unless
+ * they are marked as part of the public ABI by a bit in the Kotlin
+ * metadata annotation. Public ABI classes have to be kept verbatim, since
+ * they are copied to all call sites of a surrounding inline function.
+ *
+ * If we keep a class we will strip the bodies from all non-inline function,
+ * remove private functions, and copy inline functions verbatim. There are
+ * two corner cases to consider.
+ *
+ * First, if we have an inline function which is not callable from Java (i.e.,
+ * due to reified type parameters), we generate a private method, which nevertheless
```

Comment:
This sentence doesn't seem correct... We're currently generating public synthetic methods for public inline+reified functions. We did generate them as private, but that was before 1.3, I think